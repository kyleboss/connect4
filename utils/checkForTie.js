const checkForTie = (board) =>
  board.every((row) => row.every((cell) => cell !== null));

module.exports = checkForTie;
