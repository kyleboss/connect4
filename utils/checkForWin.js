const checkHorizontal = (player, board) => {
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row][col + 1] === player &&
        board[row][col + 2] === player &&
        board[row][col + 3] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkVertical = (player, board) => {
  for (let col = 0; col < 7; col++) {
    for (let row = 0; row < board.length - 3; row++) {
      if (
        board[row][col] === player &&
        board[row + 1][col] === player &&
        board[row + 2][col] === player &&
        board[row + 3][col] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkDiagonal = (player, board) => {
  // Check right diagonal
  for (let row = 0; row < board.length - 3; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row + 1][col + 1] === player &&
        board[row + 2][col + 2] === player &&
        board[row + 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  // Check left diagonal
  for (let row = 3; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row - 1][col + 1] === player &&
        board[row - 2][col + 2] === player &&
        board[row - 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  return false;
};

const checkForWin = (player, board) =>
  checkHorizontal(player, board) ||
  checkVertical(player, board) ||
  checkDiagonal(player, board);

// Must be CommonJS to accomodate for express server.
module.exports = checkForWin;
