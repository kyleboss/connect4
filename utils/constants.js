// Game State
const GAME_STATE_WAITING = "GAME_STATE_WAITING";
const GAME_STATE_ACTIVE = "GAME_STATE_ACTIVE";
const GAME_STATE_GAME_OVER = "GAME_STATE_GAME_OVER";

// User roles
const USER_ROLE_GUEST = "USER_ROLE_GUEST";
const USER_ROLE_RED = "USER_ROLE_RED";
const USER_ROLE_YELLOW = "USER_ROLE_YELLOW";
const USER_ROLE_NONE = "USER_ROLE_NONE";
const USER_ROLES = {
  [USER_ROLE_GUEST]: "Guest",
  [USER_ROLE_RED]: "Red",
  [USER_ROLE_YELLOW]: "Yellow",
  [USER_ROLE_NONE]: "No one",
};

// URLs
const CLIENT_PORT = 3000;
const SERVER_PORT = 3001;
const CLIENT_URL = `http://localhost:${CLIENT_PORT}`;
const SERVER_URL = `http://localhost:${SERVER_PORT}`;

// Socket Events
const SOCKET_EVENT_JOIN_GAME = "SOCKET_EVENT_JOIN_GAME";
const SOCKET_EVENT_RESET_GAME = "SOCKET_EVENT_RESET_GAME";
const SOCKET_EVENT_MAKE_MOVE = "SOCKET_EVENT_MAKE_MOVE";
const SOCKET_EVENT_ASSIGN_ROLE = "SOCKET_EVENT_ASSIGN_ROLE";
const SOCKET_EVENT_UPDATE_STATE = "SOCKET_EVENT_UPDATE_STATE";
const SOCKET_EVENT_CONNECTION = "connection"; // Official socket event.
const SOCKET_EVENT_DISCONNECT = "disconnect"; // Official socket event.

// Must be CommonJS to accomodate for express server.
module.exports = {
  GAME_STATE_WAITING,
  GAME_STATE_ACTIVE,
  GAME_STATE_GAME_OVER,
  USER_ROLE_GUEST,
  USER_ROLE_RED,
  USER_ROLE_YELLOW,
  USER_ROLE_NONE,
  USER_ROLES,
  CLIENT_PORT,
  SERVER_PORT,
  CLIENT_URL,
  SERVER_URL,
  SOCKET_EVENT_JOIN_GAME,
  SOCKET_EVENT_RESET_GAME,
  SOCKET_EVENT_MAKE_MOVE,
  SOCKET_EVENT_ASSIGN_ROLE,
  SOCKET_EVENT_UPDATE_STATE,
  SOCKET_EVENT_CONNECTION,
  SOCKET_EVENT_DISCONNECT,
};
