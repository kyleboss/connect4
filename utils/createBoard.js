const createBoard = () => Array(6).fill(Array(7).fill(null));

module.exports = createBoard;
