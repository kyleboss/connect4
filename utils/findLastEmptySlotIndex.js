const findLastEmptySlotIndex = (board, columnIndex) => {
  for (let i = board.length - 1; i >= 0; i--) {
    if (!board[i][columnIndex]) {
      return i;
    }
  }
  return -1; // Indicates the column is full
};

module.exports = findLastEmptySlotIndex;
