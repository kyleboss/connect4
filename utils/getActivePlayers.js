const { USER_ROLE_GUEST } = require("./constants");

// Active players are folks who are actually playing the game, not just watching.
// Active players are either USER_ROLE_RED or USER_ROLE_YELLOW. Since the only other
// type of player is USER_ROLE_GUEST, it's sufficient just to find the folks who are
// not guests.
const getActivePlayers = (players) =>
  Object.values(players).filter((role) => role !== USER_ROLE_GUEST);

module.exports = getActivePlayers;
