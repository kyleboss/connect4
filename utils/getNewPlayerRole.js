const {
  USER_ROLE_GUEST,
  USER_ROLE_YELLOW,
  USER_ROLE_RED,
} = require("./constants");
const getActivePlayers = require("./getActivePlayers");

// When someone connects to the game, how we decide if they are red, yellow, or a guest?
// If there are two active players already, this new person must be a guest.
// Otherwise, if there is a red user already, the new person must be yellow.
// If the above two conditions are not met: The new user is red.
const getNewPlayerRole = (players) => {
  if (getActivePlayers(players).length === 2) {
    return USER_ROLE_GUEST;
  }

  return Object.values(players).includes(USER_ROLE_RED)
    ? USER_ROLE_YELLOW
    : USER_ROLE_RED;
};

module.exports = getNewPlayerRole;
