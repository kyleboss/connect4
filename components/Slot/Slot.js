import styled, { css, keyframes } from "styled-components";

import { USER_ROLE_RED, USER_ROLE_YELLOW } from "@/utils/constants";

const dropAnimation = keyframes`
  0% { transform: translateY(-150px); }
  60% { transform: translateY(10px); }
  80% { transform: translateY(-5px); }
  100% { transform: translateY(0); }
`;

const StyledSlot = styled.div.attrs({ role: "button " })`
  position: relative;
  width: 50px;
  height: 50px;
  border: 1px solid rgb(var(--color-black));
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: ${({ disabled }) => (disabled ? "default" : "pointer")};
  background-color: rgb(var(--color-white));

  &::after {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: ${({ value }) =>
      value === USER_ROLE_RED
        ? `rgb(var(--color-red))`
        : value === USER_ROLE_YELLOW
        ? `rgb(var(--color-yellow))`
        : `rgb(var(--color-white))`};
    animation: ${({ value }) =>
      value
        ? css`
            ${dropAnimation} 0.5s ease
          `
        : "none"};
    border-radius: 50%;
  }
`;

const Slot = ({ value, onClick, disabled }) => {
  return <StyledSlot value={value} onClick={onClick} disabled={disabled} />;
};

export default Slot;
