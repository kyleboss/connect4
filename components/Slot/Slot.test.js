import React from "react";
import { render } from "@testing-library/react";

import Slot from "./Slot";
import { USER_ROLE_RED, USER_ROLE_YELLOW } from "@/utils/constants";

describe("Slot", () => {
  it("renders an empty slot correctly", () => {
    const { asFragment } = render(<Slot />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a red slot correctly", () => {
    const { asFragment } = render(<Slot value={USER_ROLE_RED} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a yellow slot correctly", () => {
    const { asFragment } = render(<Slot value={USER_ROLE_YELLOW} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a disabled slot correctly", () => {
    const { getByRole } = render(<Slot disabled />);
    const slot = getByRole("button");
    expect(slot).toHaveStyle("cursor: default");
  });
});
