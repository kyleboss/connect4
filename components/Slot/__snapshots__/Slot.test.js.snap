// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot renders a red slot correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-2b7f7c93-0 ewezXh"
    role="button "
    value="USER_ROLE_RED"
  />
</DocumentFragment>
`;

exports[`Slot renders a yellow slot correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-2b7f7c93-0 cNAEWC"
    role="button "
    value="USER_ROLE_YELLOW"
  />
</DocumentFragment>
`;

exports[`Slot renders an empty slot correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-2b7f7c93-0 bONUyQ"
    role="button "
  />
</DocumentFragment>
`;
